### PR TITLE
stat: WIP implementation of binomial distribution

### PR DESCRIPTION
**Pull request by sglyon**
NOTE: this is not 100% ready for merge.

I started working on the Binomial distribution. Most of the work is done, but there are a few issues that should be resolved before this is ready for final review

- I don't sample directly from the distribution. Instead I sample `b.N` times from the distribution `Bernoulli{P: b.P, Source: b.Source}` and count the number of successes. In theory this is correct as the Binomial represents `N` iid draws from a Bernoulli, but I think this is probably less efficient than a routine  that can sample directly from the Binomial and some of the tests in `testFullDist` fail when I use this approach.
- I have not implemented the `Quantile` function.

I didn't have the time to sort out how to sample directly from the Binomial or evaluate Quantiles, but I figured I'd share this in case members of the Gonum team either have the extra-capacity to finish this off or are willing to work with me and help me do it.

Thanks!

### CODE REVIEW

File: stat/distuv/binomial_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
```
Comment (sbinet) [inline stat/distuv/binomial_test.go]: `s/2016/2017/`

File: stat/distuv/binomial.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	bern := Bernoulli{P: b.P, Source: b.Source}
+	for i := 0; i < b.N; i++ {
+		out += bern.Rand()
+	}
+	return out
+}
+
+// Quantile returns the inverse of the cumulative probability distribution.
+func (b Binomial) Quantile(p float64) float64 {
+	// TODO(sglyon): need to implement this
```
Comment (kortschak) [inline stat/distuv/binomial.go]: http://www.lancaster.ac.uk/pg/moorhead/Work/SummerProject13.pdf

File: stat/distuv/binomial.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+// Variance returns the variance of the probability distribution.
+func (b Binomial) Variance() float64 {
+	return float64(b.N) * b.P * (1 - b.P)
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (b Binomial) Rand() float64 {
+	// TODO(sglyon): right now we just apply n independent Bernoulli...
```
Comment (kortschak) [inline stat/distuv/binomial.go]: http://www.eirene.de/Devroye.pdf may have suggestions.

File: stat/distuv/binomial.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+import (
+	"math"
+	"math/rand"
+
+	"gonum.org/v1/gonum/mathext"
+	"gonum.org/v1/gonum/stat/combin"
+)
+
+// Binomial represents a random variable that characterizes characterizes the
```
Comment (sbinet) [inline stat/distuv/binomial.go]: `s/characterizes characterizes/characterizes/`

File: stat/distuv/binomial.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	u := b.P * (1 - b.P)
+	return (1 - 6*u) / (float64(b.N) * u)
+}
+
+// LogProb computes the natural logarithm of the value of the probability density function at x.
+func (b Binomial) LogProb(x float64) float64 {
+	return math.Log(b.Prob(x))
+}
+
+// Median returns the median of the probability distribution.
```
Comment (btracey) [inline stat/distuv/binomial.go]: Are you sure this is right? The wikipedia article suggests it's not that simple.

**Comment by sglyon** [comment]
Thank you for the review. I will hopefully have another chance to take a look at this and make changes over the weekend.

File: stat/distuv/binomial.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+// Entropy returns the entropy of the distribution.
+func (b Binomial) Entropy() float64 {
+	if b.P == 0 || b.P == 1 || b.N == 0 {
+		return 0
+	}
+	p0 := 1 - b.P
+	lg := math.Log(b.P / p0)
+	lp := float64(b.N) * math.Log(p0)
+	s := math.Exp(lp) * lp
+	for k := 0; k < b.N; k++ {
```
Comment (btracey) [inline stat/distuv/binomial.go]: Needs to have an equality, `b.N` is a possible value

File: stat/distuv/binomial.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+	xf := math.Floor(x)
+	return mathext.RegIncBeta(float64(b.N)-xf, xf+1, 1-b.P)
+}
+
+// Entropy returns the entropy of the distribution.
+func (b Binomial) Entropy() float64 {
+	if b.P == 0 || b.P == 1 || b.N == 0 {
+		return 0
+	}
+	p0 := 1 - b.P
```
Comment (btracey) [inline stat/distuv/binomial.go]: Why not just

```
var ent float64
for i := 0; i < b.N; i++ {
    lp := b.LogProb(float64(i))
    ent += lp * math.Exp(lp)
}

File: stat/distuv/binomial.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	return -s
+}
+
+// ExKurtosis returns the excess kurtosis of the distribution.
+func (b Binomial) ExKurtosis() float64 {
+	u := b.P * (1 - b.P)
+	return (1 - 6*u) / (float64(b.N) * u)
+}
+
+// LogProb computes the natural logarithm of the value of the probability density function at x.
```
Comment (btracey) [inline stat/distuv/binomial.go]: `Prob` should work off of this, not the other way around.

File: stat/distuv/binomial.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+func (b Binomial) Median() float64 {
+	return math.Ceil(float64(b.N) * b.P)
+}
+
+// NumParameters returns the number of parameters in the distribution.
+func (Binomial) NumParameters() int {
+	return 2
+}
+
+// Prob computes the value of the probability distribution at x.
```
Comment (btracey) [inline stat/distuv/binomial.go]: Please replace this with `math.Exp(b.LogProb(x))`, and implement the probability in log space.

File: stat/distuv/binomial.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+	xi := int(x)
+	if float64(xi) != x {
+		return 0.0
+	}
+	return float64(combin.Binomial(b.N, xi)) *
+		math.Pow(b.P, x) *
+		math.Pow(1-b.P, float64(b.N-xi))
+}
+
+// Quantile returns the inverse of the cumulative probability distribution.
```
Comment (btracey) [inline stat/distuv/binomial.go]: We can't add this unimplemented. Please implement as per @kortschak 's suggestion or delete.

File: stat/distuv/binomial.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+func (b Binomial) Quantile(p float64) float64 {
+	// TODO(sglyon): need to implement this
+	panic("Not implemented!!!")
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (b Binomial) Rand() float64 {
+	// TODO(sglyon): right now we just apply n independent Bernoulli...
+	out := 0.0
+	bern := Bernoulli{P: b.P, Source: b.Source}
```
Comment (btracey) [inline stat/distuv/binomial.go]: Same here, if we're going to add this we should do it with good random value generation. `Rand` and `LogProb` are the most important methods, so they should be done right.

File: stat/distuv/binomial_test.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+		{
+			P: 0.9,
+			N: 10,
+		},
+		{
+			P: 0.2,
+			N: 10,
+		},
+	} {
+		testFullDist(t, dist, i, false)
```
Comment (btracey) [inline stat/distuv/binomial_test.go]: Rather than `testFullDist`, could you implement these as individual functions, like is done in `chiSquaredTest`?

**Comment by btracey** [comment]
Ping?

**Comment by sglyon** [comment]
I'm sorry to leave this as is, but I don't think I will have the time to finish the PR any time soon. My Julia implementation of the project I was considering using go/gonum for was considerably faster than my go implementation, so I am focusing efforts there for now.

Hopefully the code I started can help towards the final implementation.

Sorry if I created extra review work for little gains.

**Comment by btracey** [comment]
Happy to help with speed, and happy to have future contributions if you ever want to come back.

**Comment by sbinet** [comment]
I'd also be interested in some perf numbers and instructions to reproduce them :)

**Comment by sglyon** [comment]
Thanks for the offers to help, but be careful what you wish for! 

The binomial distribution code was a very small part of the overall project I'm using Julia for.

Knowing that, if you are still interested, let me know and I'll package up the code in a shareable way.

**Pull request closed without merging**